CMake improvements (mainly for Linux and FreeBSD)

arthurzam opened the pull request:
continue of issue #22 and PR #21 
As of the previous PR, I will add future commits as we resolve, and at the end squash everything and you will add it into the project (without merge commit).

- [x] fix finding GME without pkg-config

I added a new CMake module for finding GME, and it will even search the library and headers, thus increasing the chances to find it.
- [x] optional modules
- [x] fallback to Qt4 if no Qt5 or Qt5 < 5.6.1
- [x] FreeBSD man install

Now it should be installing to $PREFIX/man -> if you run the default prefix, as seen [here](https://www.freebsd.org/doc/en/books/porters-handbook/porting-prefix.html), then is should be in correct place.
- [x] show summary
- [ ] `make uninstall` should remove empty directories *(doc/qmplay2)*
- [x] remove QX11Extras dependency.
- [x] FFmpeg 2.2.0 or newer is needed

~~I can't manage to understand how to check what version we even use. My last resort I see is to run something like `ffmpeg -version` and then parse the output using regex.
Another question - what is the source of the minimal version - just not working? not compiling? specific "module" version nedded?~~
- [x] properly install QMPlay2-opencda.desktop

I found (finally) the proper way to use those files. But some of the concerns are: how do we distinguish KDE4 and/or KF5 in case both are installed.
- [x] install the license

I think much better name for the license file would be LICENSE instead of COPYING. What are your thoughts?
- [x] add option for ProstoPleer


~~Another question about VDPAU: in #22 you mentioned that it also needs libX11, but in the pro file there is no mention of this check. What is the correct way?
On the same file, in [FFmpeg.pro](https://github.com/zaps166/QMPlay2/blob/master/src/modules/FFmpeg/FFmpeg.pro#L45) are those sources/headers optional, if yes in what way? Are they connected to libavdevice?
Also, should we add a sub-option if USE_VDPAU is on to make optional the VDPAU without own video output?~~

- [x] add C++11 (`-std=c++11`) flags for CMake < 3.1.0 and GCC and Clang (also for GCC 6, no need to detect compiler version) and Qt 5.7.0

- [x] use CMake 2.8.11 for compatibility *(in progress, Qt4 problems)*
- [x] detect OpenGL|ES 2.0 for Qt5 and set compiler macros for it
- [x] show warning when module will be disabled due to missing dependencies

---

The full and cleaned-up task list is in #22

zaps166 commented:
> I think much better name for the license file would be LICENSE instead of COPYING. What are your thoughts?

I'm not a lawer, I can change it, sounds better :) Anyway I probably should add a license-header to every `cpp` and `hpp` file... :smiley:

> Another question about VDPAU: in #22 you mentioned that it also needs libX11, but in the pro file there is no mention of this check. What is the correct way?

I'm using `XOpenDisplay()` there, so it needs the `libX11` (it should link to it in CMake). I don't know, but it works, maybe due to ohter libraries symbols. Windows/MinGW is more restrictive in linking.

> Also, should we add a sub-option if USE_VDPAU is on to make optional the VDPAU without own video output?

I've removed macros from sources, so it needs to add the ifdefs again. They can be compiled together w/o additional option.

> I found (finally) the proper way to use those files. But some of the concerns are: how do we distinguish KDE4 and/or KF5 in case both are installed.

Arch Linux has only KF5 (and some apps from KDE4).  If KF5 is detected - copy to KF5 directory. Otherwise copy to KDE4 directory (path can differ between distros). No need to support both.

> I can't manage to understand how to check what version we even use. My last resort I see is to run something like ffmpeg -version and then parse the output using regex.

It'll fail if `ffmpeg` command tool won't be available. Probably you must check every module seaprately: libavcodec, libavformat, etc (they have different versions). Just do it as the latest thing.

> Another question - what is the source of the minimal version - just not working? not compiling? specific "module" version nedded?

FFmpeg less then 2.2.0 won't compile (e.g. `Buffer` class needs features from >= 2.2.0).

> As of the previous PR, I will add future commits as we resolve, and at the end squash everything and you will add it into the project (without merge commit).

I can squash all commits before merge, so no need to change the history of PR :)

> I added a new CMake module for finding GME, and it will even search the library and headers, thus increasing the chances to find it.

So we've got the new `cmake` dir. It will be great to have all CMake files there :smiley:, so only `*.pro` files for qmake/QtCreator will be in dirs, and entire CMake files in separate directory :)

> Now it should be installing to $PREFIX/man -> if you run the default prefix, as seen here, then is should be in correct place.

I'll check this :)

zaps166 commented:
Version merge code:
```C
#define AV_VERSION_INT(a, b, c) (a<<16 | b<<8 | c)
```
so mathematically:
- `major * 0x10000 + minor * 0x100 + micro`
- `major * 65536 + minor * 256 + micro`

---

`libavutil` version for FFmpeg 2.2.0 is:
```C
#define LIBAVUTIL_VERSION_MAJOR  52
#define LIBAVUTIL_VERSION_MINOR  66
#define LIBAVUTIL_VERSION_MICRO 100
```

`libavcodec` version for FFmpeg 2.2.0 is:
```C
#define LIBAVCODEC_VERSION_MAJOR 55
#define LIBAVCODEC_VERSION_MINOR  52
#define LIBAVCODEC_VERSION_MICRO 102
```

`libavformat` version for FFmpeg 2.2.0 is:
```C
#define LIBAVFORMAT_VERSION_MAJOR 55
#define LIBAVFORMAT_VERSION_MINOR 33
#define LIBAVFORMAT_VERSION_MICRO 100
```

`libswscale` version for FFmpeg 2.2.0 is:
```C
#define LIBSWSCALE_VERSION_MAJOR 2
#define LIBSWSCALE_VERSION_MINOR 5
#define LIBSWSCALE_VERSION_MICRO 102
```

`libswresample` version for FFmpeg 2.2.0 is:
```C
#define LIBSWRESAMPLE_VERSION_MAJOR 0
#define LIBSWRESAMPLE_VERSION_MINOR 18
#define LIBSWRESAMPLE_VERSION_MICRO 100
```

`libavresample` version for FFmpeg 2.2.0 is:
```C
#define LIBAVRESAMPLE_VERSION_MAJOR  1
#define LIBAVRESAMPLE_VERSION_MINOR  2
#define LIBAVRESAMPLE_VERSION_MICRO  0
```

arthurzam commented:
- VDPAU x11 check was added.
- Added option for Qt5X11Extras (docstring might be bad/horrible).
- Added FFmpeg versions check, according to the versions stated above.

About the KF5/KDE4: I'm afraid this will bring a big dependency tree for just one little file to install. Therefore, firstly I would like to add an option for choosing the directory. This check will just find the default one: (stop if found) search for KF5 => search for KDE4 => just set the default one (please choose). If "None" is stated by the user, we won't install it. (I'm afraid for adding dependency on the KDE big world - I prefer minimalistic system, and just for building isn't a good idea).

Also about finding packages. I can continue (like now) to just use pkg-config, but I think searching (at first we will try pkg-config) is much better for porting to new OSs. The only disadvantage is some more files in the /src/cmake/modules. If won't be hard for me.

Btw: I will update the list above, and make the text crossed when it answered (I won't delete it for future generations of bored developers). It will affect only the first message.

zaps166 commented:
Rename `FindLibGME.cmake` -> `FindGME.cmake` (looks better)
Also maybe rename `cmake_uninstall.cmake.in` -> `Uninstall.cmake.in` (it is now in `cmake` directory, so no need for 'cmake' in name)?

> About the KF5/KDE4 (...)

I don't know the default path - it depends. Try this logic ($USER_SOLID_ACTIONS_PATH is a custom patch which can be added by -DCMAKE_SOMETHING='path'):
```bash
if [ -z $USER_SOLID_ACTIONS_PATH ]; then #If user doesn't specify the path - try detect it
	if [ -d /usr/share/solid/actions ]; then
		copy to /usr/share/solid/actions
	elif [ -d /usr/share/kde4/apps/solid/actions ]; then
		copy to /usr/share/kde4/apps/solid/actions
	elif [ -d /usr/share/apps/solid/actions ]; then
		copy to /usr/share/apps/solid/actions
	else
		"Show in summary that this file won't be copied."
	fi
else
	copy to $USER_SOLID_ACTIONS_PATH
fi
```

> Also about finding packages. I can continue (like now) to just use pkg-config, but I think searching (at first we will try pkg-config) is much better for porting to new OSs. The only disadvantage is some more files in the /src/cmake/modules. If won't be hard for me.

It is less important. All Linux distros and FreeBSD have `pkg-config` (also it works on OS X and sometimes on Windows with some effort). Is it worth? Maybe later for OS X or other exotic platforms (Windows...)?

zaps166 commented:
I don't know CMake syntax, but why it has condition in `endif()` ?

https://github.com/zaps166/QMPlay2/blob/6f407c31c736b08474b646963ae7709e4e8c370c/src/cmake_uninstall.cmake.in#L5

Or in `else()`: https://github.com/zaps166/QMPlay2/blob/6f407c31c736b08474b646963ae7709e4e8c370c/src/cmake_uninstall.cmake.in#L20

? :smiley:

zaps166 commented:
> install the license

How other software handle this? Is it task for CMake or is it only task for packager?

zaps166 commented:
Added another task, see #22 (about ProstoPleer, 9aac5d112681094a3e1429896bc122d0eff73891).

arthurzam commented:
- fixed removal of QX11Extras
- added ProstoPleer
- renamed files
- fixed CMake syntax (the else and endif is a legacy thing for supporting old CMake versions, not needed in our case).

About the license file - most of them use the LICENSE file naming. It is also the default one in Arch and Gentoo.

About the searching libs - I will put it on hold. CMake has a great potential for Windows, making building as effortless for you as possible. But for now we will focus on Linux and FreeBSD, and maybe OS X.

arthurzam commented:
- solid action are installing: make dependency is only extra-cmake-modules.
- added summary
please say if the summary is OK?

zaps166 commented:
I'll start testing :)

Btw. https://github.com/zaps166/QMPlay2/pull/24/commits/1d7cb1955fb7ec102e0042071149aa3af9744dca#diff-ed2c06d232f33ec5338b0720dd0e61ebR23 (file condition)

zaps166 commented:
> added summary please say if the summary is OK?

OK!

> solid action are installing: make dependency is only extra-cmake-modules.

ECM is optional, so w/o it it just do nothing, right? It tries to push it into `/usr/local/...` (also if the prefix is `/usr`), so it doesn't auto-detect (tried on `make DESTDIR=../xxx install`).

> renamed files

`Uninstall.cmake.in` - capital letters for consistency?

> About the searching libs - I will put it on hold. CMake has a great potential for Windows, making building as effortless for you as possible. But for now we will focus on Linux and FreeBSD, and maybe OS X.

OK! Btw. look here #23

---

According to https://github.com/zaps166/QMPlay2/blob/0577c75e317bdb11c87a5aff48a73570c1d44f3b/compile_unix#L31: Use Qt5 if Qt4 is not found (also show warning if <= 5.6.0).

---

Something is wrong on FreeBSD now, cannot configure:

```sh
cmake .. -DUSE_QT5=Yes
```

```sh
CMake Error at src/gui/CMakeLists.txt:105 (install):
  install TARGETS given no RUNTIME DESTINATION for executable target
  "QMPlay2".

CMake Error at src/qmplay2/CMakeLists.txt:145 (install):
  install TARGETS given no LIBRARY DESTINATION for shared library target
  "qmplay2".
```

zaps166 commented:
> ECM is optional, so w/o it it just do nothing, right? It tries to push it into /usr/local/... (also if the prefix is /usr), so it doesn't auto-detect (tried on make DESTDIR=../xxx install).

It works, but when I use `-DCMAKE_INSTALL_PREFIX=/usr`. If I change the prefix later, then it doesn't work (e.g. in cmake-gui).

arthurzam commented:
I rebased again because of the "CMake: Enable ProstoPleer" commit (removes your change in the appropriate commit).
I think I found the problem with FreeBSD - fixed.
I rewritten the system for path of solid-action, with it's appropriate cmake module. Also it now support the value "Auto" - find the location in runtime.
Renamed the uninstall file.

ECM is optional, and if it doesn't find it, it sets the default path "/usr/share/solid/actions/" by itself. It is a recommended dependency, but not required one.

zaps166 commented:
> I think much better name for the license file would be LICENSE instead of COPYING

Done: 42c5d1c4b045f6945c9082fcf614cb484210bc06

arthurzam commented:
Never thought I will succeed, but I managed to add uninstall empty directories...
The solution is terrible (in my opinion). It traverses every file path's dir until DESTDIR and checks if it is empty.

zaps166 commented:
> added summary please say if the summary is OK?

maybe a one empty line before it :)

arthurzam commented:
Added LICENSE file
added one empty line before it

zaps166 commented:
> The solution is terrible (in my opinion). It traverses every file path's dir until DESTDIR and checks if it is empty.

I hope that the scrip doesn't remove everything :smiley:

arthurzam commented:
There are checks to find if there is no content in the directory. And only if there is no file/dir/symlink/etc in it it deletes it. I can't say it is 100% error proof, but I believe every "make uninstall" is very risky. Using the package manager to uninstall is much better.

zaps166 commented:
> The solution is terrible (in my opinion). It traverses every file path's dir until DESTDIR and checks if it is empty.

Hmm, maybe it is better to not remove empty directories, now I've got infinite loop on FreeBSD (so I don't want to try on non-VM OS :) ) - it's too dangerous. No need to remove everything, maybe after uninstalling remove only `QMPlay2` from `include` (and any other QMPlay2-only directory if exists).

zaps166 commented:
> Using the package manager to uninstall is much better

Yes, but if I install something via `make install` and it can't be uninstalled this leads to manually removing every file.

arthurzam commented:
Ok, now it will remove only /usr/share/qmplay2, /usr/include/qmplay2, /usr/lib/qmplay2 (those types).

zaps166 commented:
> Ok, now it will remove only /usr/share/qmplay2, /usr/include/qmplay2, /usr/lib/qmplay2 (those types).

Ok :)

- `/usr/lib/qmplay2/` is not removed,
- duplicated slash '/`'

```bash
-- Uninstalling directory //usr/share/qmplay2/lang//
-- Uninstalling directory //usr/share/qmplay2//
-- Uninstalling directory //usr/local/lib/qmplay2/modules//
-- Uninstalling directory //usr/local/include/QMPlay2/
```

arthurzam commented:
About duplicate slash - this is sanity check in case DESTDIR is without slash at the end (you can guess the dangerous situation). This is a small problem that saves us errors.
It remove according to the selected PREFIX when you configured CMake. the uninstall script remove exactly what you installed, and it doesn't remove anything except it (and the directories).

zaps166 commented:
> It remove according to the selected PREFIX when you configured CMake. the uninstall script remove exactly what you installed, and it doesn't remove anything except it (and the directories).

Sorry, I should write: `/usr/local/lib/qmplay2/` is not removed (still exists after `make uninstall`).

arthurzam commented:
Can you check that this directory is empty. It will only remove empty directory!

zaps166 commented:
Yes, it is empty:
```sh
$ ls -la /usr/local/lib/qmplay2/
total 259
drwxr-xr-x    2 root  wheel     2 14 cze 18:31 .
drwxr-xr-x  102 root  wheel  2519 14 cze 18:31 ..
```

arthurzam commented:
My bad! Fixed.

zaps166 commented:
> My bad! Fixed.

Ok :)

> fallback to Qt4 if no Qt5 or Qt5 < 5.6.1

```bash
CMake Error at /usr/local/share/cmake/Modules/FindQt4.cmake:1310 (message):
  Found unsuitable Qt version "5.5.1" from /usr/local/bin/qmake, this code
  requires Qt 4.x
Call Stack (most recent call first):
  CMakeLists.txt:107 (find_package)
```

Qt5 <= 5.6.0 should be selected automatically when Qt4-dev doesn't exists (https://github.com/zaps166/QMPlay2/blob/0577c75e317bdb11c87a5aff48a73570c1d44f3b/compile_unix#L31) + show the warning as in all Qt5 <= 5.6.0 cases

zaps166 commented:
Ok, all tasks are done :) Now bugfixes

zaps166 commented:
Rename `taglib` -> `TagLib` (see https://taglib.github.io/), especially for summary

---

> About duplicate slash - this is sanity check in case DESTDIR is without slash at the end (you can guess the dangerous situation). This is a small problem that saves us errors.

So missing duplicated slash for `//usr/local/include/QMPlay2/`.
But I still don't understand. `make DESTDIR=xyz install` doesn't affect uninstallation script (it still wants to remove from `CMAKE_INSTALL_PREFIX`). Even if `CMAKE_INSTALL_PREFIX` is not absolute path it still appends and prepends two slashes.
If it removes only empty dirs - it is not dangerous (IMO).

---

Installation (tested on FreeBSD) is broken. Some files goes to `CMAKE_INSTALL_PREFIX` and some files goes to `/usr` :smiley:
`bin`, `include`, `lib`, `man` - OK
`share` - Not OK (ignores `CMAKE_INSTALL_PREFIX`, always goes to `/usr`)

arthurzam commented:
About FreeBSD:
What is the right location for icons, LICENSE and CHANGELOG, mime files, etc?

There is also DESTDIR (which is by default just /). DESTDIR/PREFIX/files.

zaps166 commented:
> What is the right location for icons, LICENSE and CHANGELOG, mime files, etc?

On FreeBSD (all *BSD?) there is only one exception:
```
if (OS == "FreeBSD" && CMAKE_INSTALL_PREFIX == "/usr/local") {
    copy `man` to `CMAKE_INSTALL_PREFIX + "/man"`
} else {
    copy `man` to `CMAKE_INSTALL_PREFIX + "/share/man"`
}
```
Also described here: https://www.freebsd.org/doc/handbook/dirstructure.html

zaps166 commented:
Rename `SolidActions.cmake` -> `FindSolidActions.cmake` ?

Finding libGME works properly (tested on `game_music_emu-kode54-git`), but I don't understand this script :smiley: Does it search for directory if `pkg-config` finds it?

zaps166 commented:
The message `Problem when removing directory` is not visible when the directory can't be removed (e.g. doesn't exists).

If `$ENV{DESTDIR}${file}` doesn't begin with `/`, so for dirs can't it be just `$ENV{DESTDIR}${dir}/` ?

arthurzam commented:
Fixed

zaps166 commented:
Ok, testing...!

```
set(INSTALL_PATH_ICONS "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_DATAROOTDIR}/icons")
set(INSTALL_PATH_MIME "${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_DATAROOTDIR}/mime")
```

are common for both OS'es

zaps166 commented:
Ok, much better, thanks :)

- de-duplicate lines (https://github.com/zaps166/QMPlay2/pull/24#issuecomment-226953088),
- solid actions: `Auto` should not install it to non-existing directory if not detected,
- cannot automatically build w/o `libxv` (see #22),
- no message why the module is disabled when library/headers are missing. Only libgme shows an error: `Could NOT find LibGME (missing:  LIBGME_LIBRARY LIBGME_INCLUDE_DIR)`. User should know why the module is disabled (except libavresample, PortAudio on Linux, ALSA on non-Linux, PulseAudio on Window/OS X).

And the latest thing for uninstall: some dirs have duplicated slash (at the end of string), and some does not :) All should have one slash at the end of string.

zaps166 commented:
- license is more or less Ok, but shouldn't it be in `DOCDIR` (https://cmake.org/cmake/help/v3.0/module/GNUInstallDirs.html) together with `ChangeLog` and `Readme.md` and `TODO` files? :)
- description for solid actions: `The destination path for the solid actions, \"None\", or \"Auto\""` (fixes typo and double "or")

arthurzam commented:
All your comments were fixed.
Only messages in summary are missing.

zaps166 commented:
Ok!

> Only messages in summary are missing.

The messages can be somewhere above the summary like `libgme`

zaps166 commented:
Still cannot install w/o `sudo` to home directory:
```
CMake Error at src/gui/cmake_install.cmake:109 (file):
  file INSTALL cannot copy file
  "/home/tomo/Desktop/QMPlay2/src/gui/Unix/solid-actions/qmplay2-opencda.desktop"
  to "/usr/share/solid/actions/qmplay2-opencda.desktop".
Call Stack (most recent call first):
  src/cmake_install.cmake:37 (include)
  cmake_install.cmake:54 (include)
```
Solid actions must respect the install prefix.

---

Consistency: in `include` and `doc` directory is `QMPlay2`. In `share` and `lib` directory is `qmplay2`. Two options:
- convert in `doc` to small letters,
- convert in `share`  and `lib` to capital letters.

Which is better? And I should test whether the headers works for modules :smile:

---

Question: does `FindGME.cmake` stop when it finds `libgme` via `pkg-config` or does it search for dirs/libraries next (I don't know cmake syntax)?

zaps166 commented:
> Which is better?

- rename name in 'doc': QMPlay2 -> qmplay2 (as in my script),
- rename solid action in CMake files: qmplay2-opencda.desktop -> QMPlay2-opencda.desktop

zaps166 commented:
Solid actions: does it really use ECM?
`PC_INSTALL_PATH_SOLID_ACTION` - maybe drop ECM, check manually 6 paths (3 for `/usr` and 3 for `/usr/local/` according to my script and that's all). What the "PC_" prefix means (pkg-config?)?

zaps166 commented:
Missing GLES detection: https://github.com/zaps166/QMPlay2/blob/998a0bb699a28b8a5e4f5472950094ee4c2e8c9b/src/modules/OpenGL2/CMakeLists.txt#L29

zaps166 commented:
Qt >= 5.7.0 support for CMake < 3.1.0:
> Qt 5.7 requires C++11 support. CMake starting from version 3.1.0 implicitly adds the minimum required compiler flags (e.g. -std=gnu++11) to the Qt5 module targets. If you use CMake version older than 3.1.0, you need to add the needed -std=... to the targets linking against Qt5 modules yourself.

So need to specify `-std=c++11` for GCC (less than 6.x) and Clang.

Btw. why not using CMake 2.8.11 (for compatibility with old distros like Ubuntu 14.04 LTS or openSUSE 13.1)? According to Qt5 docs it needs only:
> It is also necessary when using an older CMake to add Qt5<Module>_EXECUTABLE_COMPILE_FLAGS to the CMAKE_CXX_FLAGS so that the -fPIC flags are added to the compile flags if necessary (as is the default with Qt 5).

Just tried with CMake 2.8.11, one warning (I removed the minimum version):
```
CMake Warning (dev) at src/gui/CMakeLists.txt:92 (add_executable):
  Policy CMP0003 should be set before this line.  Add code such as

    if(COMMAND cmake_policy)
      cmake_policy(SET CMP0003 NEW)
    endif(COMMAND cmake_policy)

  as early as possible but after the most recent call to
  cmake_minimum_required or cmake_policy(VERSION).  This warning appears
  because target "QMPlay2" links to some libraries for which the linker must
  search:

    X11, tag

  and other libraries with known full path:

    /usr/lib/libQt5Gui.so.5.6.1

  CMake is adding directories in the second list to the linker search path in
  case they are needed to find libraries from the first list (for backwards
  compatibility with CMake 2.4).  Set policy CMP0003 to OLD or NEW to enable
  or disable this behavior explicitly.  Run "cmake --help-policy CMP0003" for
  more information.
This warning is for project developers.  Use -Wno-dev to suppress it.
```
After adding `-fPIC` (`-DCMAKE_CXX_FLAGS='-fPIC'`) compiles (Qt4 and Qt5), installs and uninstalls properly (tested on current `master`) - no need to force CMake 3.0.2 :)

arthurzam commented:
I rebased.

About GME - I added printing message that will print it's finding. This is only for a test (my system uses pkg-config and it always worked!). Now it uses only pkg-config if found (before it used the information from pkg-config to find full path, like [here](https://github.com/vasi/kdelibs/blob/master/cmake/modules/FindBlueZ.cmake) from KDE.

About CMake 2.8.11 - please check on the systems you use. I have only 3.5.2 (the latest).

About OpenGL|ES - I have no idea how to check it. Many searches without any yield.

Btw. there is a big number of commits - can I squash some from the beginning? It becomes quite hard to rebase. (I must rebase to check with your latest updates).

zaps166 commented:
> About OpenGL|ES - I have no idea how to check it. Many searches without any yield.

Ok, so let's defer GLES.

> can I squash some from the beginning?

Yes, but please squash commits from yesterday only (so do 3 commits) :)

zaps166 commented:
And after rebase you should remove policy "CMP0043":
```
 Policy "CMP0043" is not known to this version of CMake.
```
btw. what the policy is?

zaps166 commented:
> if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU") 

GNU means GCC or also Clang (not tested yet)?

> if(CMAKE_VERSION VERSION_LESS "3.1")

Cmake < 3.1 and Qt version >= 5.7.0 (QMPlay2 doesn't use C++11 yet, no need to set - slows down the compilation for GCC < 6)

> set(CMAKE_CXX_FLAGS "-fPIC ${CMAKE_CXX_FLAGS}")

Qt5 docs recommends `Qt5_EXECUTABLE_COMPILE_FLAGS`. I'm not sure whether `-fPIC` is always OK, so let's use the Qt5 flags.

---

Can't use Qt4 on CMake 2.8.11 (hmm, it worked yesterday)

```
CMake Error at /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find Qt4 (missing: QT_MOC_EXECUTABLE) (found version "4.8.7")
Call Stack (most recent call first):
  /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
  /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindQt4.cmake:1375 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:123 (find_package)
```

---

Don't show message for GME path (it can be checked e.g. in cmake-gui).

arthurzam commented:
I squashed most of the start commits (stopped on the addition of make uninstall dirs).

- Added check for POLICY CMP0043. Please check.
- Changed the solid-actions check path
- removed GME path print (this was a temporary check for you).
- C++11 if CMake < 3.1 && Qt5 >= 5.7.0

zaps166 commented:
Ok, testing... :)

Btw. for Qt5 and GLESv2: `if (${Qt5Gui_OPENGL_IMPLEMENTATION} STREQUAL GLESv2)` (https://codereview.qt-project.org/#/c/79799/3/tests/auto/cmake/test_opengl_lib/CMakeLists.txt)

zaps166 commented:
Ok, so CMake 2.8.11 can't use Qt4 (works on current `master`, tested yesterday):
```
CMake Error at /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find Qt4 (missing: QT_MOC_EXECUTABLE) (found version "4.8.7")
Call Stack (most recent call first):
  /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
  /home/tomo/Desktop/cmake-2.8.11-Linux-i386/share/cmake-2.8/Modules/FindQt4.cmake:1375 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:126 (find_package)
```

Polices are OK, but I think that `cmake_policy(SET CMP0003 NEW)` is not needed (no longer complains about it if missing).

Question: `cmake_policy(SET CMP0043 NEW)` - how does it work on old CMake? Does it works properly? Maybe set it to `OLD` and fix script (I don't know)? Or it doesn't matter in QMPlay2?

arthurzam commented:
About opengles2: thank you!!! For now added for Qt5. Now searching for Qt4...
Those policies - from version to version the codebase and standards for how CMake handles information in the script changed a little. CMP0043 is needed for Qt5 defenitions (at least on my two systems).
About Qt4 - can you check before and after the [commit](https://github.com/zaps166/QMPlay2/pull/24/commits/b406a6e7ebecb0a1a0d4f4691af5ddb3ef0471f2)?

zaps166 commented:
> About Qt4 - can you check before and after the commit?

HEAD: ff45e20a9b287fe852da0b40975dc138e3ff5d82
Command (CMake 2.8.11): `./cmake /home/tomo/Desktop/QMPlay2 -DUSE_QT5=No`

Error (https://github.com/zaps166/QMPlay2/pull/24#issuecomment-226990827)

zaps166 commented:
> Those policies - from version to version the codebase and standards for how CMake handles information in the script changed a little. CMP0043 is needed for Qt5 defenitions (at least on my two systems).

Ok! But is `CMP0003` necessary?

zaps166 commented:
About GLES: https://github.com/zaps166/QMPlay2/blob/8323be5fcd5ceca40dfff1a6cc6ac323f330c9f3/src/modules/modules.pro#L5 - no way to compile with GLESv1, so show a warning and disable OpenGL module in this case

arthurzam commented:
> Ok! But is CMP0003 necessary?

Yes. CMake version 3.5.2. Without it I get this error (many times):
```
CMake Warning (dev) in src/modules/Chiptune/CMakeLists.txt:
  Policy CMP0043 is not set: Ignore COMPILE_DEFINITIONS_<Config> properties.
  Run "cmake --help-policy CMP0043" for policy details.  Use the cmake_policy
  command to set the policy and suppress this warning.
This warning is for project developers.  Use -Wno-dev to suppress it.
```
> OpenGL2 Module

So I need to disable the module if Qt4 or Qt5 with opengles1?
> Qt4 error

I mean, can you check the status before the "CMake minimal version 2.8.11"? I need to check if it part of the lower CMake version or not. It it isn't the problem then the error means CMake can't find the moc executable.

zaps166 commented:
In your warning I can see `CMP0043` policy, but I wrote about `CMP0003` :)

zaps166 commented:
> So I need to disable the module if Qt4 or Qt5 with opengles1?

Disable the module in this case **or** show error and stop cmake (user must disable it manually) - do what is easier for you (GLESv1 is very rare case) :)

zaps166 commented:
> I mean, can you check the status before the "CMake minimal version 2.8.11"? I need to check if it part of the lower CMake version or not. It it isn't the problem then the error means CMake can't find the moc executable.

If I re-run the cmake 2.8.11 it finds MOC from Qt 5.6.1 and then it can't compile.

arthurzam commented:
Ah, sorry :)
Yes it is a must one (CMP0003), because all the system of Qt5 libraries passed to the executable is using full path to lib, which this policy enables.

I ran now CMake version 3.5.2 using Qt4 (arch linux) - works properly. I even switched some times and rebuild - al working. Just to be sure - you have Qt4's moc?

zaps166 commented:
For error messages for modules, example for XVideo:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7e9bece..31db19c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -75,7 +75,8 @@ OPTION(USE_PULSEAUDIO "Build with PulseAudio support" ${LIBPULSE_FOUND})
 add_feature_info(PulseAudio USE_PULSEAUDIO "Build with PulseAudio support")
 
 if(NOT DEFINED USE_XVIDEO)
-    find_package(X11 QUIET)
+    find_package(X11)
+    find_package(X11_Xv)
     if(X11_FOUND AND X11_Xv_FOUND)
         set(DEFAULT_XVIDEO ON)
```

zaps166 commented:
> I ran now CMake version 3.5.2 using Qt4 (arch linux) - works properly. I even switched some times and rebuild - al working. Just to be sure - you have Qt4's moc?

Yes, I've got Qt4 MOC. This is about CMake 2.8.11, on 3.5.2 Qt4 works properly. Also if I tested yesterday on 2.8.11+Qt4 it worked properly on HEAD b88ae4ec4b6114b54666634d38201966aef413a7

> Yes it is a must one (CMP0003), because all the system of Qt5 libraries passed to the executable is using full path to lib, which this policy enables.

Ok!

zaps166 commented:
I must check Qt4 on CMake 2.8.11 once again like yesterday.

```
if(COMMAND cmake_policy)
    cmake_policy(SET CMP0003 NEW)
    if(POLICY CMP0043)
        cmake_policy(SET CMP0043 NEW)
    endif()
endif()
```
can it be in done as include?

zaps166 commented:
> I must check Qt4 on CMake 2.8.11 once again like yesterday.

It works properly (CMake 2.8.11 + Qt4.8.7). HEAD: b88ae4ec4b6114b54666634d38201966aef413a7, replaced: `cmake_minimum_required(VERSION 3.0.2)` -> `cmake_minimum_required(VERSION 2.8.11)` everywhere and works without any warning (also doesn't complain about `CMP0003`).

zaps166 commented:
I think `Qt5_EXECUTABLE_COMPILE_FLAGS` should include `-fPIC` automatically.

zaps166 commented:
```cmake
    find_package(ALSA QUIET)
    set(DEFAULT_ALSA ${ALSA_FOUND})
```
ALSA can be always enabled (see #22)

---

I think that you should check for all Qt5 modules (Core, Gui, Widgets, Network).

zaps166 commented:
This should fix Qt4+Cmake 2.8.11 problem:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7e9bece..54c3675 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,8 +23,7 @@ else()
 endif()
 
 find_package(Qt5Widgets QUIET)
-find_package(Qt4 QUIET)
-if(Qt5Widgets_FOUND AND (NOT Qt4_FOUND OR NOT (Qt5Widgets_VERSION VERSION_LESS 5.6.1)))
+if(Qt5Widgets_FOUND AND (Qt5Widgets_VERSION VERSION_GREATER 5.6.0))
     set(DEFAULT_QT5 ON)
 else()
     set(DEFAULT_QT5 OFF)
```

zaps166 commented:
And for compiler flags for Qt >= 5.7.0:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7e9bece..8851486 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -116,9 +116,11 @@ string(STRIP ${QMPLAY2_VERSION} QMPLAY2_VERSION)
 
 if(USE_QT5)
     find_package(Qt5Widgets REQUIRED)
-    set(Qt5_EXECUTABLE_COMPILE_FLAGS "-fPIC ${Qt5_EXECUTABLE_COMPILE_FLAGS}")
-    if(CMAKE_VERSION VERSION_LESS "3.1" AND NOT (Qt5Widgets_VERSION VERSION_LESS 5.7.0)) # cmake < 3.1 && Qt5 >= 5.7.0
-        if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" or CMAKE_CXX_COMPILER_ID STREQUAL "Clang") # gcc or clang
+    if(CMAKE_VERSION VERSION_LESS 2.8.11)
+        set(CMAKE_CXX_FLAGS "${Qt5Widgets_EXECUTABLE_COMPILE_FLAGS} ${CMAKE_CXX_FLAGS}")
+    endif()
+    if(CMAKE_VERSION VERSION_LESS 3.1 AND NOT (Qt5Widgets_VERSION VERSION_LESS 5.7.0)) # CMake < 3.1 && Qt5 >= 5.7.0
+        if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR CMAKE_CXX_COMPILER_ID STREQUAL "Clang") # GCC or Clang
             set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
         endif()
     endif()
```

arthurzam commented:
Well that a lot of replies from me :)

- `find_package(X11_Xv)` patch - not needed. This is part of what `find_package(X11)` does. It also defines all the libs and found variable.
- module detection - this (on average) 3 lines of code before every OPTION in /CMakeLists.txt is the code that selects the default state for the OPTION. I can't (!) move it anywhere else before defining the OPTION. On the other hand I can't import /src/modules before defining the OPTION block (cause it depends on it). Possible Solutions are: move this whole block into it's own file (I think the best option), move the module's options in the corresponding CMakeLists.txt (worse because maintaining is a little more tedious to find all of them in that case).
- I tried the policy code using include - unsuccessfully. (move to a new file -> include -> error as like no result from this code).
- A little about the QUIET keyword in every find_package or pkg-config use: we decided the CMake system will autodetect what the user has and what it doesn't. In case it doesn't we don't need to print the user long error log that it can't find (If you are okay with this, then I will change it). The next find with REQUIRED is for when the user selected porously to enable without the libs - then show the error. The REQUIRED should not be called at all when the USE flag is disabled.
- I added the `-fPIC` by your instructions. As you say I will do. Because I work on a rolling release systems (Arch and Gentoo) I don;t have those old versions with all this "magic". So What should I do?
- I never tried to call the build with Qt5 when for example Qt5Network is missing. With my experience on newer versions of CMake, including some module (maybe except only Core) includes all the others. And if the user is missing some component, The `qt5_use_module` should inform about it.
- About the fix for Qt4 and Cmake 2.8.11 : this removes the fallback to old Qt5 when no Qt4.
- About the fix for compiler flags for Qt >= 5.7.0 - will add them in next commit.

Thank you for creating order in this work by the tasks list in the issue. This is nearly a life saver for me. :)

zaps166 commented:
> About the fix for Qt4 and Cmake 2.8.11 : this removes the fallback to old Qt5 when no Qt4.

Oops :smiley: 

> About the fix for compiler flags for Qt >= 5.7.0 - will add them in next commit.

Ok, I've tested it - it adds properly the `-fPIC` from Qt5 flags.

> find_package(X11_Xv) patch - not needed. This is part of what find_package(X11) does. It also defines all the libs and found variable.

Maybe switch to `pkg-config` for X11?

> I tried the policy code using include - unsuccessfully. (move to a new file -> include -> error as like no result from this code).

Ok!

>  I can't (!) move it anywhere else before defining the OPTION

Ok (I don't know CMake) :)

> A little about the QUIET keyword in every find_package or pkg-config use (...)

Just remove the `QUIET` and don't append `REQUIRED` :) My proposal (not fully tested, don't know what does it break; also uses `pkg-config` for XVideo):
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7e9bece..d0a58f3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,16 +12,6 @@ include(FeatureSummary)
 find_package(PkgConfig REQUIRED)
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/src/cmake/Modules/")
 
-if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
-    find_package(ALSA QUIET)
-    set(DEFAULT_ALSA ${ALSA_FOUND})
-
-    set(DEFAULT_PORTAUDIO OFF)
-else()
-    set(DEFAULT_ALSA OFF)
-    set(DEFAULT_PORTAUDIO ON)
-endif()
-
 find_package(Qt5Widgets QUIET)
 find_package(Qt4 QUIET)
 if(Qt5Widgets_FOUND AND (NOT Qt4_FOUND OR NOT (Qt5Widgets_VERSION VERSION_LESS 5.6.1)))
@@ -30,74 +20,74 @@ else()
     set(DEFAULT_QT5 OFF)
 endif()
 
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
+    set(DEFAULT_ALSA ON)
+    set(DEFAULT_AVDEVICE ON)
+    set(DEFAULT_PORTAUDIO OFF)
+else()
+    set(DEFAULT_ALSA OFF)
+    set(DEFAULT_AVDEVICE OFF)
+    set(DEFAULT_PORTAUDIO ON)
+endif()
+
 OPTION(USE_QT5 "Build with Qt5" ${DEFAULT_QT5})
 add_feature_info(Qt5 USE_QT5 "Build with Qt5")
 OPTION(USE_PROSTOPLEER "Build with ProstoPleer support" ON)
 add_feature_info(ProstoPleer USE_PROSTOPLEER "Build with ProstoPleer support")
 OPTION(USE_AVRESAMPLE "Build with support for libavresample" OFF)
 add_feature_info(libavresample USE_AVRESAMPLE "Build with support for libavresample")
-
 OPTION(USE_ALSA "Build with ALSA support" ${DEFAULT_ALSA})
 add_feature_info(ALSA USE_ALSA "Build with ALSA support")
 OPTION(USE_OPENGL "Build with OpenGL support" ON)
 add_feature_info(OpenGL USE_OPENGL "Build with OpenGL support")
 OPTION(USE_PORTAUDIO "Build with PortAudio support" ${DEFAULT_PORTAUDIO})
 add_feature_info(PortAudio USE_PORTAUDIO "Build with PortAudio support")
-
-if(NOT DEFINED USE_TAGLIB)
-    PKG_CHECK_MODULES(TAGLIB taglib QUIET)
-endif()
-OPTION(USE_TAGLIB "Build with tags editor" ${TAGLIB_FOUND})
+OPTION(USE_TAGLIB "Build with tags editor" ON)
 add_feature_info(taglib USE_TAGLIB "Build with tags editor")
 
 if(NOT DEFINED USE_AUDIOCD)
-    PKG_CHECK_MODULES(LIBCD libcdio libcddb QUIET)
+    PKG_CHECK_MODULES(LIBCD libcdio libcddb)
 endif()
 OPTION(USE_AUDIOCD "Build with AudioCD support" ${LIBCD_FOUND})
 add_feature_info(AudioCD USE_AUDIOCD "Build with AudioCD support")
 
 if(NOT DEFINED USE_CHIPTUNE_GME)
-    find_package(GME QUIET)
+    find_package(GME)
 endif()
 OPTION(USE_CHIPTUNE_GME "Build Chiptune with GME support" ${LIBGME_FOUND})
 add_feature_info("Chiptune GME" USE_CHIPTUNE_GME "Build Chiptune with GME support")
 
 if(NOT DEFINED LIBSIDPLAYFP_FOUND)
-    PKG_CHECK_MODULES(LIBSIDPLAYFP libsidplayfp QUIET)
+    PKG_CHECK_MODULES(LIBSIDPLAYFP libsidplayfp)
 endif()
 OPTION(USE_CHIPTUNE_SID "Build Chiptune with SIDPLAY support" ${LIBSIDPLAYFP_FOUND})
 add_feature_info("Chiptune SIDPLAY" USE_CHIPTUNE_SID "Build Chiptune with SIDPLAY support")
 
 if(NOT DEFINED USE_PULSEAUDIO)
-    PKG_CHECK_MODULES(LIBPULSE libpulse-simple QUIET)
+    PKG_CHECK_MODULES(LIBPULSE libpulse-simple)
 endif()
 OPTION(USE_PULSEAUDIO "Build with PulseAudio support" ${LIBPULSE_FOUND})
 add_feature_info(PulseAudio USE_PULSEAUDIO "Build with PulseAudio support")
 
 if(NOT DEFINED USE_XVIDEO)
-    find_package(X11 QUIET)
-    if(X11_FOUND AND X11_Xv_FOUND)
-        set(DEFAULT_XVIDEO ON)
-    else()
-        set(DEFAULT_XVIDEO OFF)
-    endif()
+    PKG_CHECK_MODULES(PC_XVIDEO x11 xv)
 endif()
-OPTION(USE_XVIDEO "Build with XVideo support" ${DEFAULT_XVIDEO})
+OPTION(USE_XVIDEO "Build with XVideo support" ${PC_XVIDEO_FOUND})
 add_feature_info(XVideo USE_XVIDEO "Build with XVideo support")
 
 if(NOT DEFINED USE_FFMPEG_VAAPI)
-    PKG_CHECK_MODULES(LIBS_VAAPI libva libva-x11 QUIET)
+    PKG_CHECK_MODULES(LIBS_VAAPI libva libva-x11)
 endif()
 OPTION(USE_FFMPEG_VAAPI "Build VAAPI acceleration into FFmpeg" ${LIBS_VAAPI_FOUND})
 add_feature_info(VAAPI USE_FFMPEG_VAAPI "Build VAAPI acceleration into FFmpeg")
 
 if(NOT DEFINED USE_FFMPEG_VDPAU)
-    PKG_CHECK_MODULES(LIBS_VDPAU vdpau x11 QUIET)
+    PKG_CHECK_MODULES(LIBS_VDPAU vdpau x11)
 endif()
 OPTION(USE_FFMPEG_VDPAU "Build VDPAU acceleration into FFmpeg" ${LIBS_VDPAU_FOUND})
 add_feature_info(VDPAU USE_FFMPEG_VDPAU "Build VDPAU acceleration into FFmpeg")
 
-OPTION(USE_FFMPEG_AVDEVICE "Build FFmpeg with libavdevice suport" ON)
+OPTION(USE_FFMPEG_AVDEVICE "Build FFmpeg with libavdevice suport" ${DEFAULT_AVDEVICE})
 add_feature_info(libavdevice USE_FFMPEG_AVDEVICE "Build FFmpeg with libavdevice suport")
 
 set(LANGUAGES "All" CACHE STRING "A space-seperated list of translations to compile in to QMPlay2, \"None\", or \"All\".")
@@ -122,6 +112,9 @@ if(USE_QT5)
             set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
         endif()
     endif()
+    if(USE_QT5 AND Qt5Widgets_VERSION VERSION_LESS 5.6.1)
+        message(AUTHOR_WARNING "Qt5 >= 5.6.1 is recommended for stable usage")
+    endif()
 else()
     find_package(Qt4 REQUIRED)
     include("${QT_USE_FILE}")
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 05c3c3a..1513f60 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -3,10 +3,6 @@ cmake_minimum_required(VERSION 2.8.11)
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
 set(CMAKE_AUTOMOC ON)
 
-if(USE_QT5 AND Qt5Widgets_VERSION LESS 5.6.1)
-    message(AUTHOR_WARNING "Qt5 >= 5.6.1 is recommended for stable usage")
-endif()
-
 add_subdirectory(gui)
 add_subdirectory(qmplay2)
 add_subdirectory(modules)
diff --git a/src/gui/CMakeLists.txt b/src/gui/CMakeLists.txt
index 58c6535..8b73d07 100644
--- a/src/gui/CMakeLists.txt
+++ b/src/gui/CMakeLists.txt
@@ -130,7 +130,7 @@ if(INSTALL_PATH_SOLID_ACTION AND NOT INSTALL_PATH_SOLID_ACTION STREQUAL "None")
     if(LOCAL_INSTALL_PATH_SOLID_ACTION)
         install(FILES Unix/solid-actions/QMPlay2-opencda.desktop DESTINATION "${LOCAL_INSTALL_PATH_SOLID_ACTION}")
     else()
-        message(AUTHOR_WARNING "No path found for solid actions - won't be installed")
+        message(WARNING "No path found for solid actions - won't be installed")
     endif()
 endif()
```

zaps166 commented:
What about this:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 7e9bece..330dfa6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -23,11 +23,15 @@ else()
 endif()
 
 find_package(Qt5Widgets QUIET)
-find_package(Qt4 QUIET)
-if(Qt5Widgets_FOUND AND (NOT Qt4_FOUND OR NOT (Qt5Widgets_VERSION VERSION_LESS 5.6.1)))
+if(Qt5Widgets_FOUND AND (Qt5Widgets_VERSION VERSION_GREATER 5.6.0))
     set(DEFAULT_QT5 ON)
 else()
-    set(DEFAULT_QT5 OFF)
+    find_package(Qt4 QUIET)
+    if(NOT Qt4_FOUND AND Qt5Widgets_FOUND)
+        set(DEFAULT_QT5 ON)
+    else()
+        set(DEFAULT_QT5 OFF)
+    endif()
 endif()
 
 OPTION(USE_QT5 "Build with Qt5" ${DEFAULT_QT5})
@@ -115,7 +119,9 @@ execute_process(
 string(STRIP ${QMPLAY2_VERSION} QMPLAY2_VERSION)
 
 if(USE_QT5)
-    find_package(Qt5Widgets REQUIRED)
+    if(NOT Qt5Widgets_FOUND)
+        find_package(Qt5Widgets REQUIRED)
+    endif()
     set(Qt5_EXECUTABLE_COMPILE_FLAGS "-fPIC ${Qt5_EXECUTABLE_COMPILE_FLAGS}")
     if(CMAKE_VERSION VERSION_LESS "3.1" AND NOT (Qt5Widgets_VERSION VERSION_LESS 5.7.0)) # cmake < 3.1 && Qt5 >= 5.7.0
         if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" or CMAKE_CXX_COMPILER_ID STREQUAL "Clang") # gcc or clang
@@ -123,7 +129,9 @@ if(USE_QT5)
         endif()
     endif()
 else()
-    find_package(Qt4 REQUIRED)
+    if(NOT Qt4_FOUND)
+        find_package(Qt4 REQUIRED)
+    endif()
     include("${QT_USE_FILE}")
     add_definitions(${QT_DEFINITIONS})
 endif()
```
?

zaps166 commented:
> I never tried to call the build with Qt5 when for example Qt5Network is missing. With my experience on newer versions of CMake, including some module (maybe except only Core) includes all the others. And if the user is missing some component, The qt5_use_module should inform about it.

Ok!

arthurzam commented:
applied your patches.

zaps166 commented:
Ok!

- "taglib" -> "TagLib"
- maybe use also `pkg-config` in XVideo module?

zaps166 commented:
I don't know about GLES in Qt4, let's handle it later.

zaps166 commented:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 21e35bf..e50d2f1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,16 +12,18 @@ include(FeatureSummary)
 find_package(PkgConfig REQUIRED)
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/src/cmake/Modules/")
 
-find_package(Qt5Widgets QUIET)
-if(Qt5Widgets_FOUND AND (Qt5Widgets_VERSION VERSION_GREATER 5.6.0))
-    set(DEFAULT_QT5 ON)
-else()
-    find_package(Qt4 QUIET)
-    if(NOT Qt4_FOUND AND Qt5Widgets_FOUND)
-        set(DEFAULT_QT5 ON)
-    else()
-        set(DEFAULT_QT5 OFF)
-    endif()
+if(NOT DEFINED USE_QT5 OR USE_QT5)
+       find_package(Qt5Widgets QUIET)
+       if(Qt5Widgets_FOUND AND (Qt5Widgets_VERSION VERSION_GREATER 5.6.0))
+               set(DEFAULT_QT5 ON)
+       else()
+               find_package(Qt4 QUIET)
+               if(NOT Qt4_FOUND AND Qt5Widgets_FOUND)
+                       set(DEFAULT_QT5 ON)
+               else()
+                       set(DEFAULT_QT5 OFF)
+               endif()
+       endif()
 endif()
 
 if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
@@ -47,7 +49,7 @@ add_feature_info(OpenGL USE_OPENGL "Build with OpenGL support")
 OPTION(USE_PORTAUDIO "Build with PortAudio support" ${DEFAULT_PORTAUDIO})
 add_feature_info(PortAudio USE_PORTAUDIO "Build with PortAudio support")
 OPTION(USE_TAGLIB "Build with tags editor" ON)
-add_feature_info(taglib USE_TAGLIB "Build with tags editor")
+add_feature_info(TagLib USE_TAGLIB "Build with tags editor")
 
 if(NOT DEFINED USE_AUDIOCD)
     PKG_CHECK_MODULES(LIBCD libcdio libcddb)
@@ -112,7 +114,7 @@ if(USE_QT5)
     if(NOT Qt5Widgets_FOUND)
         find_package(Qt5Widgets REQUIRED)
     endif()
-    if(CMAKE_VERSION VERSION_LESS 2.8.11)
+    if(CMAKE_VERSION VERSION_LESS 2.8.12)
         set(CMAKE_CXX_FLAGS "${Qt5Widgets_EXECUTABLE_COMPILE_FLAGS} ${CMAKE_CXX_FLAGS}")
     endif()
     if(CMAKE_VERSION VERSION_LESS 3.1 AND NOT (Qt5Widgets_VERSION VERSION_LESS 5.7.0)) # CMake < 3.1 && Qt5 >= 5.7.0
@@ -120,7 +122,7 @@ if(USE_QT5)
             set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
         endif()
     endif()
-    if(USE_QT5 AND Qt5Widgets_VERSION VERSION_LESS 5.6.1)
+    if(Qt5Widgets_VERSION VERSION_LESS 5.6.1)
         message(AUTHOR_WARNING "Qt5 >= 5.6.1 is recommended for stable usage")
     endif()
 else()
diff --git a/src/cmake/Uninstall.cmake.in b/src/cmake/Uninstall.cmake.in
index 6a21782..3b5357e 100644
--- a/src/cmake/Uninstall.cmake.in
+++ b/src/cmake/Uninstall.cmake.in
@@ -8,6 +8,7 @@ set(DIRS_CLEAN
     "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_INCLUDEDIR@/QMPlay2/"
     "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_DATAROOTDIR@/qmplay2/lang/"
     "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_DATAROOTDIR@/qmplay2/"
+    "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_DOCDIR@"
     "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBDIR@/qmplay2/modules/"
     "@CMAKE_INSTALL_PREFIX@/@CMAKE_INSTALL_LIBDIR@/qmplay2/"
 )
@@ -30,11 +31,6 @@ foreach(file ${files})
     endif()
 endforeach()
 
-message("")
-
-list(SORT DIRS_CLEAN)
-list(REVERSE DIRS_CLEAN)
-
 foreach(dir ${DIRS_CLEAN})
     file(GLOB DIR_LIST "$ENV{DESTDIR}${dir}/*")
     list(LENGTH DIR_LIST DIR_LIST_LEN)
diff --git a/src/modules/OpenGL2/CMakeLists.txt b/src/modules/OpenGL2/CMakeLists.txt
index 1a84d14..da0c342 100644
--- a/src/modules/OpenGL2/CMakeLists.txt
+++ b/src/modules/OpenGL2/CMakeLists.txt
@@ -32,7 +32,7 @@ if(NOT USE_QT5 OR Qt5Widgets_VERSION LESS 5.6.0)
     list(APPEND OpenGL2_HDR OpenGL2OldWidget.hpp)
     list(APPEND OpenGL2_SRC OpenGL2OldWidget.cpp)
 
-    if(NOT Qt5Gui_OPENGL_IMPLEMENTATION STREQUAL GLESv2)
+    if(NOT Qt5Gui_OPENGL_IMPLEMENTATION STREQUAL GLESv2) #FIXME (see OpenGL2.pro)
         add_definitions(-DVSYNC_SETTINGS)
     endif()
 else()
@@ -41,7 +41,7 @@ else()
     list(APPEND OpenGL2_SRC OpenGL2Window.cpp OpenGL2Widget.cpp OpenGL2CommonQt5.cpp)
 endif()
 
-if((USE_QT5 AND Qt5Gui_OPENGL_IMPLEMENTATION STREQUAL GLESv2))
+if(USE_QT5 AND (Qt5Gui_OPENGL_IMPLEMENTATION STREQUAL GLESv2))
     add_definitions(-DOPENGL_ES2)
 endif()
 

```

This should be enough, please review, fix (if necessary) and apply. Then I'll merge it :)

zaps166 commented:
Hmm, what will happen if I push a commit here...?

zaps166 commented:
Oops... revert...

arthurzam commented:
Maybe revert history???
git reset

zaps166 commented:
It is merged, maybe just squash them and force push with a nice commit message?
I should learn more git :smiley:

arthurzam commented:
OK!

zaps166 commented:
I though than I can push a next commit to this PR... :cry:

arthurzam commented:
The problem is that git supports it. Not github, I think.
